shark/src/main/java/shark/internal/PathFinder.kt

```diff
@@ -462,6 +473,53 @@ internal class PathFinder(
     }
   }
 
+  private fun HeapInstance.readAllNonNullFieldsOfReferenceType(): MutableList<HeapField> {
+    // Assigning to local variable to avoid repeated lookup and cast:
+    // HeapInstance.graph casts HeapInstance.hprofGraph to HeapGraph in its getter
+    val hprofGraph = graph
+    var fieldReader: FieldValuesReader? = null
+    val result = mutableListOf<HeapField>()
+    var skipBytesCount = 0
+    for (heapClass in instanceClass.classHierarchy.toList()) {
+      for (fieldRecord in heapClass.readRecord().fields) {
+        if (fieldRecord.type != PrimitiveType.REFERENCE_HPROF_TYPE) {
+          // Skip all fields that are not references. Track how many bytes to skip
+          skipBytesCount += hprofGraph.getRecordSize(fieldRecord)
+        } else {
+          // Initialize field reader if it's not yet initialized. Replaces `lazy` without synchronization
+          if (fieldReader == null) {
+            fieldReader = hprofGraph.createFieldValuesReader(readRecord())
```

Comment:
I wrote a big comment and GitHub lost it. Damn.

Anyway: as it stands the change exposes the wrong APIs. They were internal because their design wasn't quite right. We shouldn't make them public.

Is there a way to get the same benefits as we have here, with an alternate to HeapInstance.readFields() that's more manual maybe and doesn't involve exposing all those badly internal APIs?}